### ISSUE TITLE
[FEATURE] Consul.io as service discovery and Spring cloud configuration

### ISSUE THREAD

**Issue by ogomezm** [jhipster/generator-jhipster#3608]
JHipster Registry, based in eureka, it is not that much distributed, scalable and raw performant as consul.io. I think Consul.io it is a really good option to implement a cutting edge microservice architecture.

Consul.io includes a DNS allowing a better multi-datacenter service management, more security features  and it is not limited to rest-api, it can manage database resources connection.

I think it worth it to take a look and see if it could be implemented in future releases.

**Comment by jdubois** [comment]
Thanks for explaining what Consul is :-)
Consul, like Zookeeper, is now also supported by Spring Cloud, so we should be able to support it.

However, Consul is written in Go, so it's quite obvious you can't run it inside a JHipster app. So it's much more complex to set up for everyone, from dev to prod. Of course we have Docker-ized everything in JHipster, but still a lot of people don't want (or can't!) use Docker.

People in the JHipster team ( @PierreBesson @wmarques ) are currently doing a lot of Consul, so I'll check with them in a couple of weeks to see what we can do about this.

Concerning Zookeeper, also, I like it better for several reasons:

- It's in Java, so probably it's easier to set up for us
- It's from the Apache software foundation, and I'd rather be free from any company
- This article: https://aphyr.com/posts/291-jepsen-zookeeper

I'm closing this as we won't do anything about Consul or Zookeeper in the short term, as we just don't have the resources for handling something that big, unless someone steps up and want to do that job. I'd love a good study on how we could integrate them, of course, before starting the work.

**Comment by ogomezm** [comment]
Thanks for you answer @jdubois.
I know these kind of changes are huge and maybe not worth the effort by now.

**Comment by PierreBesson** [comment]
@ogomezm I personnaly would love to see JHipster work with several Service Discovery mechanism as options.
This could actually be done by integrating [Spring Cloud Consul](http://cloud.spring.io/spring-cloud-consul/spring-cloud-consul.html). However:
* Consul is providing both Service Discovery and Distributed Configuration. it is a complete replacement for the JHipster Registry (based on Eureka and Config Server).
* As far as I know, Consul requires an agent running on each node. This is much harder to setup for our audience (Java devs).

**Comment by ogomezm** [comment]
@PierreBesson I agree with you consul would be a potential replacement of JHipster, but I prefer the approach you just mention, being able to work with several Service Discovery mechanism.

Regarding the difficulties for a (Java dev) I would say it is almost the same as if they choose to create a microservice application, there is a JHipster Registry they have to run, weather in a docker container or in their local machines, to develop the application. The difficult part it is deploying to a production environment where every node has to have the agent running, so this needs to be somehow simplified. 

Anyway I will create a Prove of Concepts for one of my current projects and I will come back to you once I have more hands-on knowledge of the challenges.

**Comment by PierreBesson** [comment]
@ogomezm Great ! Really, looking forward to seeing it.

**Comment by jdubois** [comment]
Thanks @ogomezm !! In particular I'm interested to know if this works well with Zuul -> I haven't looked at this at all
